docs/undo.md

```diff
@@ -0,0 +1,280 @@
+Transformations are a good example of that. But also point behavior: when
+a handle is implicitly being moved because it has a “smooth” connection
+to another, the opposite movement may result in a different position of
+the “implied” movement due to rounding. So in general, we need to
+efficiently copy the complete previous state of an outline to be able to
+do undo reliably.
+
+If diffing between versions can be done with text-representations of
+objects, the github.com/google/diff-match-patch library may be of
+interest; Python’s difflib is not really an option as it only does
+diffing, not patching.
+
+Perhaps much of the code can be implemented as a separate library, for
+example called “tundo”.
+
+## Considerations about resource usage
+
+Should all undo data be kept in RAM (possibly limiting the number of undo
+levels to be kept) or should undo data be flushed to disk, to minimize
+RAM usage?
+
+## Factoring
+
+From the perspective of the UI and object code, the undo API should be
+simple. Objects may have to provide efficient way to serialize themselves
+(either as binary data or as text), but the undo manager will take care
+of storing data efficiently. The undo API will have no notion about how
+efficiently or inefficiently the undo data is stored. Whether data is
+stored in memory or on disk will be completely opaque to the undo client.
+This will allow us to build a working system first, and then measure and
+optimize for performance and resource usage.
+
+## Undo vs Selection
+
+Undo is not only about the model data, but also selection data: if you do
+a step back with undo, the user also expects the selection to be the same
+as before. This notion is important because it implies undo data is more
+than just serialization of glyph data.
+
+Perhaps the tfont library should support a serialization format that is
+specialized for undo, in that it contains selection info. I need to
+figure out how this needs to be factored. Perhaps the “converter” concept
+can be used.
+
+## Undo vs Redo
+
+Each UndoManager should maintain an undo stack and a redo stack. When
+editing data, before the edit is applied, the current version of the data
+should be added to the undo stack. Performing undo is also an edit
+operation, but it’s special in that the current state of the data is now
+added to the redo stack. The design of the undo manager should make
+implementing this behavior easy.
+
+## Undo vs Notifications
+
+Depending on how notifications are done in TruFont, some leverage may be
+gained from pairing undo with the notification mechanism. A more general
+notification system may allow this more easily. (Needs more
+investigation: see the trufont.util.tracker module, and how it’s used in
+layer.paths.)
+
+## Undo Manager API concepts
+
+Flow of events when editing:
+
+- An edit action is about to happen (for example the selection will be
+  dragged)
+- The undo manager is told the name of the operation (In this case
```

Comment:
And possibly more meta data, e.g. the concerned layer, so the user knows which drag he's undoing. Maybe this info can even be used for e.g. a visual indicator effect in the UI.

Reply:
I think this would only be needed if one can edit multiple layers at the same time. If that's not the case, one can only undo whatever happened in the active layer, so it should be clear what will happen.